File: python/leetcode.127.py
import Queue
class Solution(object):
    def ladderLength1(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        if not beginWord in wordList:
            wordList.append(beginWord)

        e = 0
        if endWord in wordList:
            e = wordList.index(endWord)
        else:
            return 0
        beg = wordList.index(beginWord)
        n = len(wordList)

        mat = [[] for i in xrange(n)]

        def cmp_word(a, b):
            la = len(a)
            lb = len(b)
            if la != lb:
                return False
            hit = reduce(lambda x, y: x + y,  map(lambda (x, y): 0 if x == y else 1, zip(a, b)) )
            return hit == 1

        for i in xrange(n):
            for j in xrange(n):
                if cmp_word(wordList[i], wordList[j]):
                    mat[i].append(j)
                    mat[j].append(i)

        vis = [False] * n

        q = Queue.Queue()
        q.put((beg, 1))
        while not q.empty():
            (c, deep) = q.get()
            vis[c] = True
            if c == e:
                return deep
            for i in xrange(len(mat[c])):
                if not vis[mat[c][i]]:
                    q.put((mat[c][i], deep + 1))
        # print mat
        return 0

    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        delta = 0
        if not beginWord in wordList:
            delta = 1
            wordList.append(beginWord)

        e = 0
        if endWord in wordList:
            e = wordList.index(endWord)
        else:
            return 0
        beg = wordList.index(beginWord)
        n = len(wordList)

        mat = [[] for i in xrange(n)]

        def cmp_word(a, b):
            la = len(a)
            lb = len(b)
            if la != lb:
                return False
            hit = reduce(lambda x, y: x + y,  map(lambda (x, y): 0 if x == y else 1, zip(a, b)) )
            return hit == 1

        for i in xrange(n):
            for j in xrange(n):
                if cmp_word(wordList[i], wordList[j]):
                    mat[i].append(j)
                    mat[j].append(i)

        vis = [0] * n

        # Search from beg to e

        q1 = Queue.Queue()
        q2 = Queue.Queue()
        q1.put((beg, 1))
        q2.put((e, -1))
        vis[beg] = 1
        vis[e] = -1

        def sgn(x):
            if x == 0: return x
            return x / abs(x)

        def bfs(q, flag):
            c, d = q.get()
            for i in xrange(len(mat[c])):
                if sgn(vis[mat[c][i]]) != flag:
                    print "{} {}: {} -> {}".format("q1" if flag > 0 else "q2", d, wordList[c], wordList[mat[c][i]])
                    if sgn(vis[mat[c][i]]) == -flag:
                        return True, abs(d - vis[mat[c][i]])
                    else:
                        vis[mat[c][i]] = d + flag
                        q.put((mat[c][i], d + flag))
            return False, 0

        while (not q1.empty()) or (not q2.empty()):
            flag = 0
            if not q1.empty():
                ans, d = bfs(q1, 1)
                if ans:
                    return d
            if not q2.empty():
                ans, d = bfs(q2, -1)
                if ans:
                    return d

        # print mat
        return 0


sln = Solution()
# 0 5 5 0 2 5 4 4
# print sln.ladderLength("hot", "dog", ["hot", "dog"]) # 0
print sln.ladderLength("hit", "cog", ["hot","dot","dog","lot","log","cog"]) # 5
# print sln.ladderLength("hit", "cog", ["hot","dot","dog","cog"]) # 5
# print sln.ladderLength("hit", "cog", ["hot","dot","dog","lot","log"]) # 0
# print sln.ladderLength("a", "c", ["a", "b", "c"]) # 2
# print sln.ladderLength("kiss", "tusk", ["miss","dusk","kiss","musk","tusk","diss","disk","sang","ties","muss"]) # 5
# print sln.ladderLength("cat", "fin", ["cat", "fat", "fin", "fit"]) # 4
# print sln.ladderLength("cat", "fin", ["ion","rev","che","ind","lie","wis","oct","ham","jag","ray","nun","ref","wig","jul","ken","mit","eel","paw","per","ola","pat","old","maj","ell","irk","ivy","beg","fan","rap","sun","yak","sat","fit","tom","fin","bug","can","hes","col","pep","tug","ump","arc","fee","lee","ohs","eli","nay","raw","lot","mat","egg","cat","pol","fat","joe","pis","dot","jaw","hat","roe","ada","mac"]) # 4

File: cpp/leetcode.127.cpp
#include <queue>
#include <vector>
#include <map>
#include <algorithm>
using namespace std;

vector<int> mat[10000];
int vis[10000];

int sgn(int x){
    if(x == 0) return 0;
    if (x > 0) return 1;
    return -1;
}

class Solution {
public:
    bool cmp_word(const string & a, const string & b){
        if (a.size() != b.size()) return false;
        int hit = 0;
        for(int i = 0; i < a.size(); i++){
            if(a[i] != b[i]) hit ++;
        }
        return (hit == 1);
    }
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        if (find(wordList.begin(), wordList.end(), beginWord) == wordList.end()){
            wordList.push_back(beginWord);
        }
        int beg, e;

        beg = find(wordList.begin(), wordList.end(), beginWord) - wordList.begin();
        auto ee = find(wordList.begin(), wordList.end(), endWord);
        if (ee == wordList.end()){
            return 0;
        }
        e = ee - wordList.begin();
        int n = wordList.size();

        for(int i = 0; i < n; i++)
            mat[i].clear();

        for(int i = 0; i < n; i++)
            for(int j = 0; j < n; j++)
                if(cmp_word(wordList[i], wordList[j])){
                    mat[i].push_back(j);
                    // mat[j].push_back(i);
                }

        memset(vis, 0, sizeof vis);

        queue<pair<int, int>> q1;
        queue<pair<int, int>> q2;
        q1.push(make_pair(beg, 1));
        q2.push(make_pair(e, -1));
        vis[beg] = 1;
        vis[e] = -1;

        auto bfs = [&](queue<pair<int, int>> & q, int flag){
            int c, d;
            tie(c, d) = q.front();
            q.pop();
            for(int i = 0; i < mat[c].size(); i++){
                if(sgn(vis[mat[c][i]]) != flag){
                    // printf( "%s %d: %s -> %s\n", flag > 0 ? "q1" : "q2", d, wordList[c].c_str(), wordList[mat[c][i]].c_str());
                    if (sgn(vis[mat[c][i]]) == -flag){
                        return make_pair(true, abs(d - vis[mat[c][i]]));
                    }else{
                        vis[mat[c][i]] = d + flag;
                        q.push(make_pair(mat[c][i], d + flag));
                    }
                }
            }
            return make_pair(false, 0);
        };

        while ((! q1.empty()) || (! q2.empty())){
            if (! q1.empty()){
                bool ans; int dd;
                tie(ans, dd) = bfs(q1, 1);
                if (ans)
                    return dd;
            }
            if (! q2.empty()){
                bool ans; int dd;
                tie(ans, dd) = bfs(q2, -1);
                if (ans)
                    return dd;
            }
        }
        return 0;
    }
};
